provisioner/clusterpy.go

```diff
@@ -328,70 +324,74 @@ func (p *clusterpyProvisioner) Provision(ctx context.Context, logger *log.Entry,
 	}
 
 	// provision node pools
-	nodePoolProvisioner := &AWSNodePoolProvisioner{
-		awsAdapter:      awsAdapter,
-		instanceTypes:   instanceTypes,
-		nodePoolManager: nodePoolManager,
-		bucketName:      bucketName,
-		config:          channelConfig,
-		cluster:         cluster,
-		azInfo:          azInfo,
-		logger:          logger,
+	caNodePoolProvisioner := &AWSNodePoolProvisioner{
+		NodePoolTemplateRenderer: NodePoolTemplateRenderer{
+			awsAdapter:     awsAdapter,
+			config:         channelConfig,
+			cluster:        cluster,
+			bucketName:     bucketName,
+			logger:         logger,
+			encodeUserData: true,
+		},
+		instanceTypes: instanceTypes,
+		azInfo:        azInfo,
+	}
+
+	karpenterProvisioner := &KarpenterNodePoolProvisioner{
+		NodePoolTemplateRenderer: NodePoolTemplateRenderer{
+			awsAdapter:     awsAdapter,
+			config:         channelConfig,
+			cluster:        cluster,
+			bucketName:     bucketName,
+			logger:         logger,
+			encodeUserData: false,
+		},
+		execManager: p.execManager,
+		tokenSource: p.tokenSource,
 	}
 
 	// group node pools based on their profile e.g. master
 	nodePoolGroups := groupNodePools(
 		logger,
 		cluster,
+		caNodePoolProvisioner,
+		karpenterProvisioner,
 	)
 
-	for _, g := range nodePoolGroups {
-		err := nodePoolProvisioner.Provision(ctx, g.NodePools, values)
-		if err != nil {
-			return err
-		}
+	err = nodePoolGroups["masters"].provisionNodePoolGroup(ctx, values, updater, cluster, p.applyOnly)
+	if err != nil {
+		return err
+	}
 
-		// custom function that checks if the node pools are "ready"
-		err = g.ReadyFn()
-		if err != nil {
-			return err
-		}
+	if err = ctx.Err(); err != nil {
+		return err
+	}
 
-		if err = ctx.Err(); err != nil {
-			return err
-		}
+	err = p.apply(ctx, logger, cluster, deletions, manifests)
```

Comment:
This changes the update sequece from `master->workers->manifests` to `master->manifests->workers`.

Is there a way to keep the old flow?

Reply:
`karpenter worker pools (provisioners)` depends on `karpenter CRDs & karpenter validation webhook a container in the karpenter controller).`
I tried this: `master -> apply necessary karpenter manifests only -> workers -> manifests`, this did not work as karpenter controller has dependency on other manifests. so I had to bring all the manifests before the worker nodes

as an alternative, we can experiment creating the karpenter provisioners workers before their dependencies (injecting them to etcd cluster directly to bypass api-server validation) or other ways.

I thought we discussed this and agreed upon this approach of `master->manifests->workers`

Reply:
Being this close to Cyber Week I would be worried about introducing this change now as it would apply to ALL clusters also where we don't test karpenter. Can we somehow do it via a config item so this order is only the case when karpenter support is enabled. Allowing us to move forward with testing karpenter in test, but stay without changes to behavior in production.